File: packages/cli-go/internal/config/network_v2_test.go
package config

import (
	"bytes"
	"encoding/json"
	"os"
	"testing"
)

func TestVersionedNetworkSchemaAndGolden(t *testing.T) {
	raw, err := os.ReadFile("../../../../schemas/config/v2/schema.json")
	if err != nil || !bytes.Equal(raw, schemaV2) {
		t.Fatal("embedded v2 schema differs", err)
	}
	raw, err = os.ReadFile("../../../../schemas/fixtures/config/v2/vectors.json")
	if err != nil {
		t.Fatal(err)
	}
	var vector struct{ Canonical, SHA256 string }
	if json.Unmarshal(raw, &vector) != nil {
		t.Fatal("invalid vector")
	}
	d, err := Normalize([]byte(vector.Canonical))
	if err != nil || d.Normalized != vector.Canonical || d.Hash != vector.SHA256 {
		t.Fatal("v2 independent canonical/hash differs", err)
	}
	// Submission carries the validated explicit version to the matching HTTP boundary.
	var submission map[string]any
	if json.Unmarshal([]byte(vector.Canonical), &submission) != nil {
		t.Fatal("fixture invalid")
	}
	submission["release"] = map[string]any{"mode": "test-only", "targets": []any{}}
	input, err := json.Marshal(submission)
	if err != nil {
		t.Fatal(err)
	}
	if _, err = Normalize(input); err != nil {
		t.Fatal("valid test-only v2 refused by validator", err)
	}
	if parsed, err := Parse(input); err != nil || parsed.SchemaVersion != 2 {
		t.Fatal("valid v2 must retain its submission version", err)
	}
	for _, tc := range []struct {
		key   string
		value any
	}{
		{"mode", "unrestricted"}, {"permissions", []any{}}, {"maximumConnections", 0},
		{"maximumConnections", 1.5}, {"maximumBytesPerSecond", uint64(4294967296)},
		{"maximumConnections", true}, {"resolver", "customer.example"},
	} {
		var value map[string]any
		if json.Unmarshal([]byte(vector.Canonical), &value) != nil {
			t.Fatal("fixture invalid")
		}
		value["network"].(map[string]any)[tc.key] = tc.value
		raw, err := json.Marshal(value)
		if err != nil {
			t.Fatal(err)
		}
		if _, err = Normalize(raw); err == nil {
			t.Fatalf("accepted invalid v2 %s", tc.key)
		}
	}
	for _, host := range []string{"localhost", "127.0.0.1", "0x7f.0.0.1", "0177.0.0.1", "*.example", "A.example", "a.example\n"} {
		var value map[string]any
		if json.Unmarshal([]byte(vector.Canonical), &value) != nil {
			t.Fatal("fixture invalid")
		}
		value["network"].(map[string]any)["permissions"].([]any)[0].(map[string]any)["hostname"] = host
		raw, err := json.Marshal(value)
		if err != nil {
			t.Fatal(err)
		}
		if _, err = Normalize(raw); err == nil {
			t.Fatal("accepted invalid hostname")
		}
	}
}

File: packages/config-schema/test/network-v2.test.mjs
import assert from "node:assert/strict";
import { readFileSync } from "node:fs";
import test from "node:test";
import {
  ConfigurationError,
  hashConfiguration,
  normalizeConfiguration,
  parseConfiguration,
  validateConfiguration,
} from "../dist/index.js";

const vectors = JSON.parse(
  readFileSync(
    new URL(
      "../../../schemas/fixtures/config/v2/vectors.json",
      import.meta.url,
    ),
  ),
);
const fresh = () => JSON.parse(vectors.canonical);

test("explicit configuration v2 reproduces frozen canonical bytes and hash", () => {
  const v = fresh();
  assert.equal(normalizeConfiguration(v), vectors.canonical);
  assert.equal(hashConfiguration(v), vectors.sha256);
  assert.deepEqual(parseConfiguration(vectors.canonical), v);
  const reordered = fresh();
  reordered.network.permissions.reverse();
  assert.doesNotThrow(() => validateConfiguration(reordered));
  assert.notEqual(
    hashConfiguration(reordered),
    vectors.sha256,
    "request order is not silently rewritten",
  );
});

test("versioned request never changes legacy none or accepts a relabelled legacy network", () => {
  const legacy = JSON.parse(
    readFileSync(
      new URL(
        "../../../schemas/fixtures/config/valid/hosted.normalized.json",
        import.meta.url,
      ),
    ),
  );
  const original = hashConfiguration(legacy);
  const enabled = fresh();
  enabled.apiVersion = "provenance.dev/v1";
  assert.throws(() => validateConfiguration(enabled), ConfigurationError);
  legacy.apiVersion = "provenance.dev/v2";
  assert.throws(() => validateConfiguration(legacy), ConfigurationError);
  legacy.network = {
    mode: "none",
    permissions: [],
    maximumConnections: 0,
    maximumBytesPerSecond: 0,
  };
  assert.doesNotThrow(() => validateConfiguration(legacy));
  assert.notEqual(hashConfiguration(legacy), original);
  legacy.network.maximumConnections = 1;
  assert.throws(() => validateConfiguration(legacy), ConfigurationError);
});

test("v2 requires explicit finite tuples and rejects unknown authority inputs", () => {
  const invalid = [
    ["mode", "unrestricted"],
    ["mode", "None"],
    ["permissions", []],
    ["maximumConnections", 0],
    ["maximumConnections", 4294967296],
    ["maximumConnections", 1.5],
    ["maximumBytesPerSecond", 0],
    ["maximumBytesPerSecond", -1],
    ["maximumBytesPerSecond", "65536"],
    ["resolver", "customer.example"],
    ["namespace", "/host/path"],
    ["allowlist", []],
  ];
  for (const [key, value] of invalid) {
    const v = fresh();
    v.network[key] = value;
    assert.throws(() => validateConfiguration(v), ConfigurationError, key);
  }
  for (const key of Object.keys(fresh().network)) {
    const v = fresh();
    delete v.network[key];
    assert.throws(() => validateConfiguration(v), ConfigurationError, key);
  }
  for (const patch of [
    { hostname: "A.example" },
    { hostname: "a.example\n" },
    { hostname: "127.0.0.1" },
    { hostname: "0x7f.0.0.1" },
    { hostname: "0177.0.0.1" },
    { hostname: "*.example" },
    { hostname: "a.example." },
    { hostname: "é.example" },
    { hostname: "localhost" },
    { hostname: "a".repeat(64) + ".example" },
    { hostname: "-a.example" },
    ...[0, 25, 53, 465, 587, 853, 65536].map((port) => ({ port })),
    { transport: "icmp" },
    { transport: "TCP" },
    { transport: 1 },
    { ports: [443] },
  ]) {
    const v = fresh();
    Object.assign(v.network.permissions[0], patch);
    assert.throws(
      () => validateConfiguration(v),
      ConfigurationError,
      JSON.stringify(patch),
    );
  }
  const duplicate = fresh();
  duplicate.network.permissions.push(duplicate.network.permissions[0]);
  assert.throws(() => validateConfiguration(duplicate), ConfigurationError);
  const bounded = fresh();
  bounded.network.permissions = Array.from({ length: 128 }, (_, i) => ({
    hostname: "a.example",
    port: 1000 + i,
    transport: "tcp",
  }));
  bounded.network.maximumConnections =
    bounded.network.maximumBytesPerSecond = 4294967295;
  assert.doesNotThrow(() => validateConfiguration(bounded));
  bounded.network.permissions.push({
    hostname: "b.example",
    port: 443,
    transport: "udp",
  });
  assert.throws(() => validateConfiguration(bounded), ConfigurationError);
  const future = fresh();
  future.apiVersion = "provenance.dev/v3";
  assert.throws(() => validateConfiguration(future), ConfigurationError);
});
